com.unity.ml-agents/Runtime/Sensors/StackingSensor.cs

```diff
@@ -67,11 +63,40 @@ public StackingSensor(ISensor wrapped, int numStackedObservations)
             }
 
             // TODO support arbitrary stacking dimension
-            m_Shape[0] *= numStackedObservations;
-            m_StackedObservations = new float[numStackedObservations][];
-            for (var i = 0; i < numStackedObservations; i++)
+            m_Shape[m_Shape.Length - 1] *= numStackedObservations;
+
+            if (m_WrappedSensor.GetCompressionType() == SensorCompressionType.None)
             {
-                m_StackedObservations[i] = new float[m_UnstackedObservationSize];
+                m_StackedObservations = new float[numStackedObservations][];
+                for (var i = 0; i < numStackedObservations; i++)
+                {
+                    m_StackedObservations[i] = new float[m_UnstackedObservationSize];
+                }
+            }
+            else
+            {
+                m_StackedCompressedObservations = new byte[numStackedObservations][];
+                m_EmptyCompressedObservation = CreateEmptyPNG();
+                for (var i = 0; i < numStackedObservations; i++)
+                {
+                    m_StackedCompressedObservations[i] = m_EmptyCompressedObservation;
+                }
+
+                // Construct compression mapping
+                var compressibleSensor = m_WrappedSensor as ICompressibleSensor;
+                var wrappedMapping = compressibleSensor.GetCompressionMapping();
+                int wrappedNumChannel = shape[2];
+                int wrappedMapLength = wrappedMapping.Length;
+                int offset;
+                m_CompressionMapping = new int[wrappedMapLength * m_NumStackedObservations];
+                for (var i = 0; i < numStackedObservations; i++)
+                {
+                    offset = wrappedNumChannel * i;
+                    for (var j = 0; j < wrappedMapLength; j++)
+                    {
+                        m_CompressionMapping[j + wrappedMapLength * i] = wrappedMapping[j] >= 0 ? wrappedMapping[j] + offset : -1;
```

Comment:
I think it should be the responsibility of the StackingSensor to pad the wrapped sensor's mapping to a multiple of 3, not the wrapped sensor.

Reply:
If we don't do this in GridSensor we cannot use GridSensor on its own without stacking right?

Reply:
Not sure I understand. GridSensor works fine without stacking now.

If the wrapped sensor has a compression mapping of size N, the stacking sensor should have a one of size 
`(numStacks) * 3 * ((N + 2) / 3)`.

Reply:
Yes GridSensor is fine now without mapping. But I mean the wrapped sensor should return the padded mapping otherwise it will be problematic without stacking (N itself should be multiple of 3). 
Because when using the sensor on its own, GetCompressionMapping() should return a mapping that's the same size as the compressed observation, which is padded. 
Also the stacked compressed observation is constructed by stacking the padded observation from wrapped sensor, and the mapping should correspond with that. 

But I agree it would be better to double check this in StackingSensor.

Reply:
> But I mean the wrapped sensor should return the padded mapping otherwise it will be problematic without stacking (N itself should be multiple of 3).

I disagree. If there are only 4 channels in the observation, then `[0, 1, 2, 3]` is an OK mapping for the wrapped sensor. It's up to the StackingSensor to pad this to `[0, 1, 2, 3, -1, -1, 4, 5, 6, 7, -1, -1 ...]`

Reply:
The mapping is not only used by stacking sensor but can also be sent to python directly. In that case GridSensor will send [0, 1, 2, 3] to python with an observation of 6 channels which can be very confusing of whether it's an error or the rest should be dropped.

Reply:
After second thoughts I decided to not send the mapping at all for GridSensor/CameraSensor and have the python code handle that in the old way.

My original idea is to make them all have the same interface so that they can all be handled in a clean and simple way, but if we're allowing sensors that doesn't follow this interface then it's becoming uglier than just keep it the old way and handle them differently.

By doing this I'm keeping the strong assumption that channel=1 is always equivalent to a mapping of {0, 0, 0} and all others has an identical mapping (non-stacking case).

Please let me know if there's still any backward-compat problem.

Reply:
Yeah, this sounds like a better approach.